scripts/ingest_folder.py

```diff
@@ -17,29 +18,70 @@
     action=argparse.BooleanOptionalAction,
     default=False,
 )
+parser.add_argument(
+    "--log-file",
+    help="Optional path to a log file. If provided, logs will be written to this file.",
+    type=str,
+    default=None,
+)
 args = parser.parse_args()
 
+# Set up logging to a file if a path is provided
+if args.log_file:
+    file_handler = logging.FileHandler(args.log_file, mode="a")
+    file_handler.setFormatter(
+        logging.Formatter(
+            "[%(asctime)s.%(msecs)03d] [%(levelname)s] %(message)s",
+            datefmt="%Y-%m-%d %H:%M:%S",
+        )
+    )
+    logger.addHandler(file_handler)
+
+
+total_documents = 0
+current_document_count = 0
+
+
+def count_documents(folder_path: Path) -> None:
+    global total_documents
+    for file_path in folder_path.iterdir():
+        if file_path.is_file():
+            total_documents += 1
+        elif file_path.is_dir():
+            count_documents(file_path)
+
 
 def _recursive_ingest_folder(folder_path: Path) -> None:
+    global current_document_count, total_documents
     for file_path in folder_path.iterdir():
         if file_path.is_file():
+            current_document_count += 1
+            progress_msg = f"Document {current_document_count} of {total_documents} ({(current_document_count / total_documents) * 100:.2f}%)"
+            logger.info(progress_msg)
             _do_ingest(file_path)
         elif file_path.is_dir():
             _recursive_ingest_folder(file_path)
 
 
 def _do_ingest(changed_path: Path) -> None:
-    if changed_path.exists():
-        print(f"\nIngesting {changed_path}")
-        ingest_service.ingest(changed_path.name, changed_path)
+    try:
+        if changed_path.exists():
+            logger.info(f"Started ingesting {changed_path}")
+            ingest_service.ingest(changed_path.name, changed_path)
+            logger.info(f"Completed ingesting {changed_path}")
+    except Exception as e:
+        logger.error(f"Failed to ingest document: {changed_path}. Error: {e}")
```

Comment:
Using `exception` instead of `error` is [adding the stack trace after the log](https://docs.python.org/3/library/logging.html#logging.Logger.exception) - I don't know if you want it, or if you just want the string representation of the error, so putting it as a suggestion:

```suggestion
    except Exception as e:
        logger.exception(f"Failed to ingest document: {changed_path}. Error: {e}")
```

Reply:
As I don't know what the common practice is, I don't have a personal opinion on it. 

The only point coming to my mind which potentially speaks against it, is that the structure of the log file will be interrupted, which might make parsing it more difficult.
After all, commonly JSON objects are logged for this reason, as they are still one entry, one line, but allow to include a stack trace for example.

I am happy to do the change if it is deemed as the better option.

Reply:
> After all, commonly JSON objects are logged for this reason, as they are still one entry, one line, but allow to include a stack trace for example.

I agree

> The only point coming to my mind which potentially speaks against it, is that the structure of the log file will be interrupted, which might make parsing it more difficult.

If programmatic parsing is wanted, then indeed, it might not be the best solution, but if this file has vocation to be read by humans, I'd suggest using `exception`, as it keeps (and display) the full error instead of the error message only (making troubleshooting way easier)